python/src/equistore/operations/sum_samples.py

```diff
@@ -0,0 +1,94 @@
+    :param block: -> input block
+    :param sample_names: -> names of samples to sum
+    """
+    for sample in sample_names:
+        assert (
+            sample in block.samples.names
+        ), "the values in sample_names should be contained in block.samples.names"
+
+    # get the indeces of the selected sample
+    sample_selected = [block.samples.names.index(sample) for sample in sample_names]
+    # reshaping the samples in a 2D array
+    samples = block.samples.view(dtype=np.int32).reshape(block.samples.shape[0], -1)
+    s, index = np.unique(samples[:, sample_selected], return_inverse=True, axis=0)
+    values_result = _dispatch.zeros_like(
+        block.values, shape=(s.shape[0],) + block.values.shape[1:]
+    )
+    _dispatch.index_add(values_result, block.values, index)
+
+    result_block = TensorBlock(
+        values=values_result,
+        samples=Labels(
+            [sample for sample in sample_names],
+            np.array([i for i in s], dtype=np.int32),
+        ),
+        components=block.components,
+        properties=block.properties,
+    )
+
+    for parameter, gradient in block.gradients():
+        # here we need to copy because we want to modify the samples array
+        samples = (
+            gradient.samples.view(dtype=np.int32)
+            .reshape(gradient.samples.shape[0], -1)
+            .copy()
+        )
+
+        samples[:, 0] = index[samples[:, 0]]
+
+        s_gradient, index_gradient = np.unique(
+            samples[:, :], return_inverse=True, axis=0
+        )
+        data_result = _dispatch.zeros_like(
+            gradient.data, shape=(s_gradient.shape[0],) + gradient.data.shape[1:]
+        )
+        _dispatch.index_add(data_result, gradient.data, index_gradient)
+
+        result_block.add_gradient(
+            parameter,
+            data_result,
+            Labels(
+                ["sample"] + [i for i in gradient.samples.names[1:]],
+                np.array([i for i in s_gradient], dtype=np.int32),
+            ),
+            gradient.components,
+        )
+
+    return result_block
+
+
+def sum_samples(tensor: TensorMap, sample_names: List[str]) -> TensorMap:
+    """Create a new :py:class:`TensorMap` with the same keys as
+    as the input `tensor`, and each :py:class:`TensorBlock`
+    is obtained summing the corresponding input :py:class:`TensorBlock`
+    over the rows with the same sample_names.
+
+    For example if sample_name = ["structure"] it sum over
+    all the sample with the same structure
+    if sample_name = ["structure", "center"] it sums over all the rows with
+    the same values for the "structure" and the "center"
+
+    :param tensor:  input :py:class:`TensorMap`
+    :param sample_names: names of samples to sum
+    """
+
+    blocks = []
+    for key, block in tensor:
+        for sample in sample_names:
+            if sample not in block.samples.names:
+                raise ValueError(f"TensorBlock with key={key} has not sample={sample}")
```

Comment:
Since the sample names are the same for all blocks, you can move this check outside of the block loop